### PR TITLE
Dataset namespacing

### PR DESCRIPTION
**Pull request by inglesp**

### CODE REVIEW

File: openprescribing/gcutils/bigquery.py [hunk 1]
```diff
@@ -349,3 +391,33 @@ def results_to_dicts(results):
...
     return [gcbq.SchemaField(*field) for field in fields]
+
+
+class SafeDict(dict):
+    def __missing__(self, key):
+        return '{' + key + '}'
+
+
+def interpolate_sql(sql, **substitutions):
+    '''Interpolates substitutions (plus datasets) into given SQL.
```
Comment (sebbacon) [inline openprescribing/gcutils/bigquery.py]: "plus datasets defined by the `DATASETS` constant"?

File: openprescribing/gcutils/bigquery.py [hunk 2]
```diff
@@ -349,3 +391,33 @@ def results_to_dicts(results):
...
+
+    Many of our SQL queries contain template variables, because the names of
+    certain tables or fields are generated at runtime, and because each test
+    run uses different dataset names.  This function replaces template
+    variables with the corresponding values in substitutions, or with the
+    dataset name.
+
+    >>> interpolate_sql('SELECT {col} from {hscic}.table', col='c')
```
Comment (sebbacon) [inline openprescribing/gcutils/bigquery.py]: >>> DATASETS = {'hscic': 'hscic_12345'}
>>> interpolate_sql('SELECT {col} from {hscic}.table', col='c')

File: openprescribing/gcutils/bigquery.py [hunk 3]
```diff
@@ -17,25 +18,41 @@
...
+
+try:
+    DATASETS['test'] = settings.BQ_TEST_DATASET
+except AttributeError:
+    pass
+
+
 class Client(object):
-    def __init__(self, dataset_name=None):
+    def __init__(self, dataset_key=None):
```
Comment (sebbacon) [inline openprescribing/gcutils/bigquery.py]: Maybe docstring here defining purpose/meaning of `dataset_key`?

**Comment by inglesp** [comment]
This is ready for merging.  The Code Climate warning is spurious.

**Pull request merged**